### PR TITLE
ARROW-10640: [C++] A "where" kernel to combine two arrays based on a mask

### PR DESCRIPTION
**Pull request by nirandaperera**
Adding a preliminary impl for an `if_else(cond: Datum, left: Datum, right: Datum)` function. It works as follows,
```python
def if_else(cond, left, right):
    for c, true_val, false_val in zip(cond, left, right):
        if c:
            yield true_val
        else:
            yield false_val
```
`null` values will be promoted to the output. 

Current commit only provides only Array-Array-Array kernels for the function.

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    ::arrow::internal::BitmapAnd(out_validity->data(), 0, cond.buffers[0]->data(),
+                                 cond.offset, len, 0, out_validity->mutable_data());
+  }
+
+  output->buffers[0] = std::move(out_validity);
+  output->GetNullCount();  // update null count
+  return Status::OK();
+}
+
+template <typename Type, bool swap = false, typename Enable = void>
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_if_else.cc:73]: What is `swap` doing?

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+using internal::BitBlockCount;
+using internal::BitBlockCounter;
+
+namespace compute {
+
+namespace {
+
+// nulls will be promoted as follows
+// cond.val && (cond.data && left.val || ~cond.data && right.val)
+Status promote_nulls(KernelContext* ctx, const ArrayData& cond, const ArrayData& left,
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_if_else.cc:34]: nit: use CamelCase (`PromoteNulls`)

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+  static Status Call(KernelContext* ctx, const Scalar& cond, const Scalar& left,
+                     const Scalar& right, Scalar* out) {
+    return Status::OK();
+  }
+};
+
+template <typename Type>
+struct ResolveExec {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch.length == 0) return Status::OK();
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_if_else.cc:194]: Do you really need to check for this? You can assume you have three arguments.

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+  static Status Call(KernelContext* ctx, const Scalar& cond, const Scalar& left,
+                     const Scalar& right, Scalar* out) {
+    // todo impl
+    return Status::OK();
+  }
+};
+
+template <typename Type, bool swap>
+struct IfElseFunctor<Type, swap, enable_if_t<is_boolean_type<Type>::value>> {
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_if_else.cc:135]: nit: `type_traits.h` already has an `enable_if_boolean`

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:135]: (same with `enable_if_number` etc)

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    std::memcpy(out_values, right_data, right.length * sizeof(T));
+
+    const auto* cond_data = cond.buffers[1]->data();  // this is a BoolArray
+    BitBlockCounter bit_counter(cond_data, cond.offset, cond.length);
+
+    // selectively copy values from left data
+    const T* left_data = left.GetValues<T>(1);
+    int64_t offset = cond.offset;
+
+    // todo this can be improved by intrinsics. ex: _mm*_mask_store_e* (vmovdqa*)
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_if_else.cc:99]: I would assume memcpy already does this for you.

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        } else {  // SSS
+          return IfElseFunctor<Type>::Call(ctx, *batch[0].scalar(), *batch[1].scalar(),
+                                           *batch[2].scalar(), out->scalar().get());
+        }
+      }
+    }
+  }
+};
+
+void AddPrimitiveKernels(const std::shared_ptr<ScalarFunction>& scalar_function,
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_if_else.cc:249]: A lot of these function names are a little generic, can we rename them to reflect that they're for IfElse? We might end up putting other similar kernels here.

**Review by lidavidm — COMMENTED** [review]
Just some quick initial feedback. I think the overall approach looks fine. Testing this thoroughly will be important.

For a different PR I added a `DatumFromJSON` to make it easier to test functions that accept a lot of permutations of scalar/array arguments. I could pull that out separately if that would make testing easier here. (That way you don't need a lot of overloads and boilerplate to test all the possible input combinations.)

**Comment by nirandaperera** [comment]
> Just some quick initial feedback. I think the overall approach looks fine. Testing this thoroughly will be important.
> 
> For a different PR I added a `DatumFromJSON` to make it easier to test functions that accept a lot of permutations of scalar/array arguments. I could pull that out separately if that would make testing easier here. (That way you don't need a lot of overloads and boilerplate to test all the possible input combinations.)

That would be great @lidavidm . If you could send me the PR# that would be great! :-)

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc (hunk 1 above)
Comment (nirandaperera) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:73]: My idea is to reuse the impl for the cases like, `cond, left: Array, right: Scalar` and `cond, left: Sca;ar, right: Array`. In the second scenario, I can swap left and right and invert the cond without changing the loop mechanism.

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc (hunk 5 above)
Comment (nirandaperera) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:99]: Do you mean, load with mask?

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:99]: Sorry, I think I misunderstood the optimization you're thinking about. How would SIMD help here?

**Comment by lidavidm** [comment]
> > Just some quick initial feedback. I think the overall approach looks fine. Testing this thoroughly will be important.
> > For a different PR I added a `DatumFromJSON` to make it easier to test functions that accept a lot of permutations of scalar/array arguments. I could pull that out separately if that would make testing easier here. (That way you don't need a lot of overloads and boilerplate to test all the possible input combinations.)
> 
> That would be great @lidavidm . If you could send me the PR# that would be great! :-)

See #10386

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc (hunk 5 above)
Comment (nirandaperera) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:99]: Say, you first copy `right` to `ouput`. Then, `cond` becomes a mask to store `left` onto `output`. For that there are specialized SIMD instructions. 
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#cats=Store&text=mask_store&expand=5564,5566

So, we can drop the `BitBlock` objects, and remove all the loops and memcpy inside the `while` loop. We'd have to handle the alignment though.

**Comment by nirandaperera** [comment]
I am closing this PR because there are some major refactors and it would be better to review them fresh.

**Pull request closed without merging**